### PR TITLE
Fix for WFCORE-2765. CLI copy directories when adding modules.

### PR DESCRIPTION
**Pull request by jfdenise**
evolve module add to copy directories or create empty directory if resource doesn't exist.

### CODE REVIEW

**Comment by jfdenise** [comment]
retest this please!

File: cli/src/main/java/org/jboss/as/cli/handlers/module/ASModuleHandler.java [hunk 1]
```diff
@@ -359,10 +369,14 @@ protected void addModule(CommandContext ctx, final ParsedCommandLine parsedCmd)
...
+        boolean allowNonExisting = allowNonExistingResources.isPresent(parsedCmd);
         for(int i = 0; i < resourceArr.length; ++i) {
             final File f = new File(pathCompleter.translatePath(resourceArr[i]));
-            if(!f.exists()) {
-                throw new CommandLineException("Failed to locate " + f.getAbsolutePath());
+            if (!f.exists() && !allowNonExisting) {
+                throw new CommandLineException("Failed to locate " + f.getAbsolutePath()
+                        + ", if you defined non-existing resource on purpose you should "
+                        + "use " + allowNonExistingResources.getFullName() + " option");
             }
```
Comment (bstansberry) [inline cli/src/main/java/org/jboss/as/cli/handlers/module/ASModuleHandler.java]: s/non-existing/a non-existing/g
s/use/use the/g

File: cli/src/main/java/org/jboss/as/cli/handlers/module/ASModuleHandler.java [hunk 2]
```diff
@@ -571,6 +585,96 @@ public static XMLExtendedStreamWriter create(XMLStreamWriter writer) throws Comm
...
+                    Path symTarget = Files.readSymbolicLink(file);
+                    File symTargetFile = symTarget.toFile();
+                    if (!symTargetFile.isAbsolute()) {
+                        if (file.getParent() != null) {
+                            symTarget = file.getParent().resolve(symTarget);
+                        } else {
+                            throw new IOException("Recursive symbolic link: "
+                                    + file.toFile().getAbsolutePath() + "=>"
+                                    + symTargetFile.getCanonicalPath()
+                                    + "can't copy directory");
```
Comment (bstansberry) [inline cli/src/main/java/org/jboss/as/cli/handlers/module/ASModuleHandler.java]: Needs a space before "can't"

Comment (bstansberry) [inline cli/src/main/java/org/jboss/as/cli/handlers/module/ASModuleHandler.java]: Actually, I think the format at L662 looks better.

File: cli/src/main/resources/help/module.txt [hunk 3]
```diff
@@ -58,6 +58,10 @@ ARGUMENTS
                    (usually jar files) separated by a filesystem-specific
                    path separator, i.e. java.io.File.pathSeparatorChar.
                    The file(s) specified will be copied to the created module's directory.
+                   If the specified path is a directory, it will be recursively copied to
+                   the module directory. If the directory is non existing, an empty directory
```
Comment (bstansberry) [inline cli/src/main/resources/help/module.txt]: s/is non existing/does not exist/g

File: cli/src/main/resources/help/module.txt [hunk 4]
```diff
@@ -107,3 +111,6 @@ ARGUMENTS
                    NOTE: this argument makes sense only when the module.xml file
                    is generated, i.e. when the --module-xml argument isn't
                    specified.
+
+--allow-non-existing-resources - Used in conjunction with --resources in order to create
```
Comment (bstansberry) [inline cli/src/main/resources/help/module.txt]: --allow-nonexistent-resources is better. "Nonexistent" is an ordinary English word.

File: cli/src/main/resources/help/module.txt [hunk 5]
```diff
@@ -107,3 +111,6 @@ ARGUMENTS
...
+                                 empty directories for non existing referenced resources.
```
Comment (bstansberry) [inline cli/src/main/resources/help/module.txt]: empty directories for referenced resources that do not exist.

File: cli/src/main/java/org/jboss/as/cli/handlers/module/ASModuleHandler.java [hunk 6]
```diff
@@ -571,6 +585,96 @@ public static XMLExtendedStreamWriter create(XMLStreamWriter writer) throws Comm
...
+            copyDirectory(source, target);
+        } catch (IOException ex) {
+            Exception removalException = null;
+            try {
+                // Remove fully the module.
+                handler.removeModule(ctx.getParsedCommandLine(), ctx);
+            } catch (Exception ee) {
+                removalException = ee;
+            }
+            String msg = "An error occured while copying directory  "
```
Comment (bstansberry) [inline cli/src/main/java/org/jboss/as/cli/handlers/module/ASModuleHandler.java]: s/occured/occurred

**Pull request merged**